File: lms_fe/app/_services/customer/customer.ts
import axios from "@/app/_axios/axios";
import { CustomerData, CustomerEducation, CustomerEmployment, defaultCustomerData, titleName } from "@/app/models/customer";

interface fetchEducationsResp {
    code: number;
    message: string;
    data: {
        Education: {
        eid: string;
        th_name: string;
        en_name: string;
        details: string;
        created_date: string;
        updated_date: string;
        }[];
    };
}

interface fetchEmploymentsResp {
    code: number;
    message: string;
    data: {
        Employment: {
            emid: string;
            th_name: string;
            en_name: string;
            details: string;
            created_date: string;
            updated_date: string;
        }[];
    };
}

interface GetCustomerDashboardDetailResp {
    total: number
    data: CustomerData[]
}

interface fetchCustomersResp {
    code: number;
    message: string;
    data: {
        customers: {
            cid: string;
            th_name: string;
            th_lastname: string;
            id_card: string;
            passport_no: string;
            mobile_no: string;
            status: string;
            created_date: string;
        }[],
        total: number;
    };
}


export const fetchCustomers = async (rowsPerPage: number, page: number):Promise<GetCustomerDashboardDetailResp> => {
    
    const requestBody = {
        limit: rowsPerPage,
        offset: rowsPerPage * page , 
    }

    try {
        const response = await axios.post<fetchCustomersResp>(
            '/customer/get-all',
            requestBody,
        );

        if (response.status === 200 || response.status === 201) {
            const custs: CustomerData[] = response.data.data.customers.map((c) => ({
                ...defaultCustomerData,
                id: c.cid,
                information: {
                    ...defaultCustomerData.information,
                    thName: c.th_name,
                    thLastname: c.th_lastname,
                    idcard: c.id_card,
                    passportNo: c.passport_no,
                },
                address: {
                    ...defaultCustomerData.address,
                    mobilePhone: c.mobile_no,
                },
                status: c.status,
            }));

            return Promise.resolve({
                total: response.data.data.total,
                data: custs
            })
        } else {
            return Promise.reject({})
        }
    } catch(e) {
        console.error(e)
        return Promise.reject({})
    }
};



interface fetchCustomerResp {
    code: number;
    message: string;
    data: {
        cid: string,
        title: {
            tid: string,
            th_name: string,
            en_name: string,
            created_date: string,
            updated_date: string,
        },
        th_name: string,
        en_name: string,
        th_lastname: string,
        en_lastname: string,
        ic_card: string,
        passport_no: string,
        address: {
            contact_name: string,
            building_no: string,
            building_name: string,
            floor: string,
            room_no: string,
            village_no: string,
            alley: string,
            street: string,
            sub_district: string,
            district: string,
            province: string,
            postal_code: string,
            home_phone: string,
        },
        mobile_no: string,
        email: string,
        education: {
            eid: string,
            th_name: string,
            en_name: string,
            details: string,
            created_date: string,
            updated_date: string,
        },
        employment: {
            emid: string,
            th_name: string,
            en_name: string,
            details: string,
            created_date: string,
            updated_date: string,
        },
        salary: number,
        status: string,
        other_income: number,
    };
}
export const fetchCustomer = async(cid: string): Promise<CustomerData> => {
    const resp = await axios.post<fetchCustomerResp>("/customer/view", {cid: cid})
    return {
        id: resp.data.data.cid,
        createdDate: "",
        information: {
            title: {
                ...defaultCustomerData.information.title,
                tid: resp.data.data.title.tid,
                thName: resp.data.data.title.th_name ,
                enName: resp.data.data.title.en_name,
            },
            thName: resp.data.data.th_name,
            thLastname: resp.data.data.th_lastname,
            enName: resp.data.data.en_name,
            enLastname: resp.data.data.en_lastname,
            idcard: resp.data.data.ic_card,
            passportNo: resp.data.data.passport_no,
        },
        address: {
            contactName: resp.data.data.address.contact_name,
            houseNumber: resp.data.data.address.building_no,
            building: resp.data.data.address.building_name,
            floor: resp.data.data.address.floor,
            roomNumber: resp.data.data.address.room_no,
            village: resp.data.data.address.village_no,
            alley: resp.data.data.address.alley,
            street: resp.data.data.address.street,
            subDistrict: resp.data.data.address.sub_district,
            district: resp.data.data.address.district,
            province: resp.data.data.address.province,
            postalCode: resp.data.data.address.postal_code,
            homePhone: resp.data.data.address.home_phone,
            mobilePhone: resp.data.data.mobile_no,
            email: resp.data.data.email,
        },
        education: {
            ...defaultCustomerData.education,
            eid: resp.data.data.education.eid,
            thName: resp.data.data.education.th_name,
            enName: resp.data.data.education.en_name,
            details: resp.data.data.education.details,
        },
        employment: {
            ...defaultCustomerData.employment,
            emid: resp.data.data.employment.emid,
            thName: resp.data.data.employment.th_name,
            enName: resp.data.data.employment.en_name,
            details: resp.data.data.employment.details,
        },
        income: {
            salary: resp.data.data.salary,
            other: resp.data.data.other_income,
        },
        status: resp.data.data.status,
    }
}


export const fetchEducations = async (): Promise<CustomerEducation[]> => {
    try {
        const resp = await axios.get<fetchEducationsResp>("/education/");
        if (resp.status === 200) {
            const edus: CustomerEducation[] = resp.data.data.Education.map((e) => ({
                eid: e.eid,
                thName: e.th_name,
                enName: e.en_name,
                details: e.details,
                createdDate: e.created_date,
                updatedDate: e.updated_date,
            }));
            return Promise.resolve(edus);
        } else {
            return Promise.reject({});
        }
    } catch (e) {
        console.error(e)
        return Promise.reject({});
    }
};

export const fetchEmployments = async (): Promise<CustomerEmployment[]> => {
    
    try {
        const response = await axios.get<fetchEmploymentsResp>("/employment/");
        if (response.status === 200 ) {
            const emps: CustomerEmployment[] = response.data.data.Employment.map((e) => ({
                emid: e.emid,
                thName: e.th_name,
                enName: e.en_name,
                details: e.details,
                createdDate: e.created_date,
                updatedDate: e.updated_date,
            }));
            return Promise.resolve(emps)
        } else {
            return Promise.reject({})
        }
    } catch (e) {
        console.error(e)
        return Promise.reject({})
    }
};


type newCustomerReq = {
    cid?: string;
    title_id: string;
    th_name: string;
    en_name: string;
    th_lastname: string;
    en_lastname: string;
    ic_card: string;
    passport_no: string;
    address: {
        contact_name: string;
        building_no: string;
        building_name: string;
        floor: string;
        room_no: string;
        village_no: string;
        alley: string;
        street: string;
        sub_district: string;
        district: string;
        province: string;
        postal_code: string;
        home_phone: string;
    };
    mobile_no: string;
    email: string;
    education_id: string;
    employment_id: string;
    salary: number;
    other_income: number;
}

type newCustomerResp = {
    code: number
    message: string
    data: {
        cid: string
    }
}

export const newCustomer = async (req: newCustomerReq): Promise<newCustomerResp> => {
    const resp = await axios.post<newCustomerResp>("/customer/new", req)
    return {
        code: resp.data.code,
        message: resp.data.message,
        data: resp.data.data
    }
}

export function convertCustomerDataToRequestBody(customerData: CustomerData, cid?: string): newCustomerReq {
    return {
        cid: cid,
        title_id: customerData.information.title.tid,
        th_name: customerData.information.thName,
        en_name: customerData.information.enName,
        th_lastname: customerData.information.thLastname,
        en_lastname: customerData.information.enLastname,
        ic_card: customerData.information.idcard,
        passport_no: customerData.information.passportNo,
        address: {
            contact_name: customerData.address.contactName,
            building_no: customerData.address.houseNumber,
            building_name: customerData.address.building,
            floor: customerData.address.floor,
            room_no: customerData.address.roomNumber,
            village_no: customerData.address.village,
            alley: customerData.address.alley,
            street: customerData.address.street,
            sub_district: customerData.address.subDistrict,
            district: customerData.address.district,
            province: customerData.address.province,
            postal_code: customerData.address.postalCode,
            home_phone: customerData.address.homePhone,
        },
        mobile_no: customerData.address.mobilePhone,
        email: customerData.address.email,
        education_id: customerData.education.eid, // assuming `eid` is numeric
        employment_id: customerData.employment.emid, // assuming `emid` is numeric
        salary: customerData.income.salary,
        other_income: customerData.income.other,
    };
}

interface fetchCustomerTitleResp {
    code: number;
    message: string;
    data: {
        Title: {
            tid: string;
            th_name: string;
            en_name: string;
            created_date: string;
            updated_date: string;
        }[]
    };
}

export const fetchCustomerTitles = async (): Promise<titleName[]> => {
    const response = await axios.get<fetchCustomerTitleResp>("/customer-title/");
    const tit: titleName[] = response.data.data.Title.map((t) => ({
        tid: t.tid,
        thName: t.th_name,
        enName:  t.en_name,
        createdDate: t.created_date,
        updatedDate: t.updated_date,
    }));
    return tit
};

export const removeCustomer = async (cid: string): Promise<string> =>{
    interface Resp {
        code: number;
        message: string;
        data: {
            cid: string;
        };
    }
    const response = await axios.post<Resp>(
        '/customer/remove',
        {
            cid: cid
        },
    );
    return response.data.data.cid
}

File: lms_be/internal/services/customer.go
package services

import (
	"fmt"
	"lms_be/internal/errors"
	"lms_be/internal/models"
	"lms_be/internal/ports"
	"lms_be/pkg"
	dateformat "lms_be/pkg/date_format"

	"github.com/google/uuid"
)

type customerServices struct {
	repo              ports.CustomerRepositories
	addrRepo          ports.AddressRepositories
	customerTitleRepo ports.CustomerTitleRepositories
}

func NewCustomerServices(
	repo ports.CustomerRepositories,
	addrRepo ports.AddressRepositories,
	customerTitleRepo ports.CustomerTitleRepositories,
) *customerServices {
	return &customerServices{
		repo:              repo,
		addrRepo:          addrRepo,
		customerTitleRepo: customerTitleRepo,
	}
}

func (c *customerServices) New(req models.NewCustomerReq) (*models.NewCustomerResp, *models.ServiceError) {
	if errMsg := pkg.RequestValidator(req); errMsg != "" {
		errResp := errors.ErrBadRequest
		errResp.Message = errMsg
		return nil, &errResp
	}

	cid, _ := uuid.NewRandom()
	customerRepo := models.CustomerRepo{
		Cid:          cid.String(),
		TitleId:      req.TitleId,
		THName:       req.THName,
		ENName:       &req.ENName,
		THLastname:   req.THLastname,
		ENLastname:   &req.ENLastname,
		IdCard:       req.IdCard,
		PassportNo:   &req.PassportNo,
		MobileNo:     req.MobileNo,
		Email:        &req.Email,
		EducationId:  req.EducationId,
		EmploymentId: req.EmploymentId,
		Salary:       &req.Salary,
		Status:       "ACTIVE",
		OtherIncome:  &req.OtherIncome,
	}
	rows, err := c.repo.New(customerRepo)

	if rows <= 0 || err != nil {
		fmt.Println(err)
		return nil, &errors.ErrNewCustomer
	}

	aid, _ := uuid.NewRandom()
	rows, err = c.addrRepo.New(models.AddressRepo{
		Aid:          aid.String(),
		Cid:          cid.String(),
		ContactName:  &req.Address.ContactName,
		BuildingNo:   req.Address.BuildingNo,
		BuildingName: &req.Address.BuildingName,
		Floor:        &req.Address.Floor,
		RoomNo:       &req.Address.RoomNo,
		VillageNo:    &req.Address.VillageNo,
		Alley:        &req.Address.Alley,
		Street:       &req.Address.Street,
		Subdistrict:  req.Address.Subdistrict,
		District:     req.Address.District,
		Province:     req.Address.Province,
		PostalCode:   req.Address.PostalCode,
		HomePhone:    &req.Address.HomePhone,
	})

	if rows <= 0 || err != nil {
		fmt.Println(err)

		// Rollback create customer
		rows, err := c.repo.Remove(customerRepo)
		if rows <= 0 || err != nil {
			fmt.Println(err)
			return nil, &errors.ErrRollBackCustomer
		}

		return nil, &errors.ErrNewAddress
	}

	return &models.NewCustomerResp{Cid: cid.String()}, nil
}

func (c *customerServices) View(req models.ViewCustomerReq) (*models.ViewCustomerResp, *models.ServiceError) {
	if errMsg := pkg.RequestValidator(req); errMsg != "" {
		errResp := errors.ErrBadRequest
		errResp.Message = errMsg
		return nil, &errResp
	}

	customer, err := c.repo.FindById(req.Cid)
	if customer == nil || err != nil {
		fmt.Println(err)
		return nil, &errors.ErrCustomerNotFound
	}

	customerWithAll, err := c.repo.FindWithAllById(req.Cid)
	if customerWithAll == nil || err != nil {
		fmt.Println(err)
		return nil, &errors.ErrCustomerNotFound
	}

	return &models.ViewCustomerResp{
		Cid: req.Cid,
		Customer: models.Customer{
			Title: models.CustomerTitle{
				Tid:         customerWithAll.CustomerTitleRepo.Tid,
				THName:      customerWithAll.CustomerTitleRepo.THName,
				ENName:      customerWithAll.CustomerTitleRepo.ENName,
				CreatedDate: customerWithAll.CustomerTitleRepo.CreatedDate.Format(dateformat.DDMMYYYYHHmmss),
				UpdatedDate: customerWithAll.CustomerTitleRepo.UpdatedDate.Format(dateformat.DDMMYYYYHHmmss),
			},
			THName:     customerWithAll.CustomerRepo.THName,
			ENName:     pkg.DerefString(customerWithAll.CustomerRepo.ENName),
			THLastname: customerWithAll.CustomerRepo.THLastname,
			ENLastname: pkg.DerefString(customerWithAll.CustomerRepo.ENLastname),
			IdCard:     customerWithAll.CustomerRepo.IdCard,
			PassportNo: pkg.DerefString(customerWithAll.CustomerRepo.PassportNo),
			Address: models.Address{
				ContactName:  pkg.DerefString(customerWithAll.AddressRepo.ContactName),
				BuildingNo:   customerWithAll.AddressRepo.BuildingNo,
				BuildingName: pkg.DerefString(customerWithAll.AddressRepo.BuildingName),
				Floor:        pkg.DerefString(customerWithAll.AddressRepo.Floor),
				RoomNo:       pkg.DerefString(customerWithAll.AddressRepo.RoomNo),
				VillageNo:    pkg.DerefString(customerWithAll.AddressRepo.VillageNo),
				Alley:        pkg.DerefString(customerWithAll.AddressRepo.Alley),
				Street:       pkg.DerefString(customerWithAll.AddressRepo.Street),
				Subdistrict:  customerWithAll.AddressRepo.Subdistrict,
				District:     customerWithAll.AddressRepo.District,
				Province:     customerWithAll.AddressRepo.Province,
				PostalCode:   customerWithAll.AddressRepo.PostalCode,
				HomePhone:    pkg.DerefString(customerWithAll.AddressRepo.HomePhone),
			},
			MobileNo: customerWithAll.CustomerRepo.MobileNo,
			Email:    pkg.DerefString(customerWithAll.CustomerRepo.Email),
			Education: models.Education{
				Eid:         customerWithAll.EducationRepo.Eid,
				THName:      customerWithAll.EducationRepo.THName,
				ENName:      customerWithAll.EducationRepo.ENName,
				Details:     customerWithAll.EducationRepo.Details,
				CreatedDate: customerWithAll.EducationRepo.CreatedDate.Format(dateformat.DDMMYYYYHHmmss),
				UpdatedDate: customerWithAll.EducationRepo.UpdatedDate.Format(dateformat.DDMMYYYYHHmmss),
			},
			Employment: models.Employment{
				Emid:        customerWithAll.EmploymentRepo.Emid,
				THName:      customerWithAll.EmploymentRepo.THName,
				ENName:      customerWithAll.EmploymentRepo.ENName,
				Details:     customerWithAll.EmploymentRepo.Details,
				CreatedDate: customerWithAll.EmploymentRepo.CreatedDate.Format(dateformat.DDMMYYYYHHmmss),
				UpdatedDate: customerWithAll.EmploymentRepo.UpdatedDate.Format(dateformat.DDMMYYYYHHmmss),
			},
			Salary:      pkg.DerefFloat64(customerWithAll.CustomerRepo.Salary),
			Status:      customerWithAll.Status,
			OtherIncome: pkg.DerefFloat64(customerWithAll.CustomerRepo.OtherIncome),
		},
	}, nil
}

func (c *customerServices) Edit(req models.EditCustomerReq) (*models.EditCustomerResp, *models.ServiceError) {
	if errMsg := pkg.RequestValidator(req); errMsg != "" {
		errResp := errors.ErrBadRequest
		errResp.Message = errMsg
		return nil, &errResp
	}

	customer, err := c.repo.FindById(req.Cid)
	if customer == nil || err != nil {
		fmt.Println(err)
		return nil, &errors.ErrCustomerNotFound
	}

	rows, err := c.repo.Update(models.CustomerRepo{
		Cid:          req.Cid,
		TitleId:      req.TitleId,
		THName:       req.THName,
		ENName:       &req.ENName,
		THLastname:   req.THLastname,
		ENLastname:   &req.ENLastname,
		IdCard:       req.IdCard,
		PassportNo:   &req.PassportNo,
		MobileNo:     req.MobileNo,
		Email:        &req.Email,
		EducationId:  req.EducationId,
		EmploymentId: req.EmploymentId,
		Salary:       &req.Salary,
		OtherIncome:  &req.OtherIncome,
	})
	if rows <= 0 || err != nil {
		fmt.Println(err)
		return nil, &errors.ErrUpdateCustomer
	}

	address, err := c.addrRepo.FindByCid(req.Cid)
	if address == nil || err != nil {
		fmt.Println(err)
		return nil, &errors.ErrAddressNotFound
	}

	rows, err = c.addrRepo.Update(models.AddressRepo{
		Aid:          address.Aid,
		Cid:          req.Cid,
		ContactName:  &req.Address.ContactName,
		BuildingNo:   req.Address.BuildingNo,
		BuildingName: &req.Address.BuildingName,
		Floor:        &req.Address.Floor,
		RoomNo:       &req.Address.RoomNo,
		VillageNo:    &req.Address.VillageNo,
		Alley:        &req.Address.Alley,
		Street:       &req.Address.Street,
		Subdistrict:  req.Address.Subdistrict,
		District:     req.Address.District,
		Province:     req.Address.Province,
		PostalCode:   req.Address.PostalCode,
		HomePhone:    &req.Address.HomePhone,
	})
	if rows <= 0 || err != nil {
		fmt.Println(err)

		// Rollback customer
		rows, err := c.repo.Update(*customer)
		if rows <= 0 || err != nil {
			return nil, &errors.ErrRollBackCustomer
		}

		return nil, &errors.ErrUpdateAddress
	}

	return &models.EditCustomerResp{
		Cid: req.Cid,
	}, nil
}

func (c *customerServices) Remove(req models.RemoveCustomerReq) (*models.RemoveCustomerResp, *models.ServiceError) {
	if errMsg := pkg.RequestValidator(req); errMsg != "" {
		errResp := errors.ErrBadRequest
		errResp.Message = errMsg
		return nil, &errResp
	}

	// TODO: Add finding customer before remove customer becuase if we can not remove address we must rollback by insert customer again
	rows, err := c.repo.Remove(models.CustomerRepo{Cid: req.Cid})
	if rows <= 0 || err != nil {
		fmt.Println(err)
		return nil, &errors.ErrRemoveCustomer
	}

	rows, err = c.addrRepo.Remove(models.AddressRepo{Cid: req.Cid})
	if rows <= 0 || err != nil {
		fmt.Println(err)
		return nil, &errors.ErrRemoveAddress
	}

	return &models.RemoveCustomerResp{Cid: req.Cid}, nil
}

func (c *customerServices) GetAll(req models.GetAllCustomerReq) (*models.GetAllCustomerResp, *models.ServiceError) {

	if req.Limit == 0 {
		req.Limit = 100
	}

	queryResult, err := c.repo.Find(req.Limit, req.Offset)
	if err != nil {
		return nil, &errors.ErrCustomerNotFound
	}

	customers := []models.GetAllCustomerDetailResp{}
	for _, cust := range queryResult {
		customers = append(customers, models.GetAllCustomerDetailResp{
			Cid:         cust.Cid,
			THName:      cust.THName,
			THLastname:  cust.THLastname,
			IdCard:      cust.IdCard,
			PassportNo:  cust.PassportNo,
			MobileNo:    cust.MobileNo,
			Status:      cust.Status,
			CreatedDate: cust.CreatedDate.Format(dateformat.DDMMYYYYHHmmss),
		})
	}

	total, _ := c.repo.Count()

	return &models.GetAllCustomerResp{
		Customers: customers,
		Total:     total,
	}, nil
}
